rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -238,129 +218,124 @@ int main(int argc, char **argv){
     );
 
 
+
+    auto complex_taps = filters.get_filter_taps();
+
     DSPCore *dp = new DSPCore(&dsp_to_brian_begin, &dsp_to_brian_end, &dsp_to_data_write,
                              sig_options, sp_packet.sequence_num(), rx_rate, output_sample_rate,
-                             rx_freqs, &filters, beam_phases,
+                             rx_freqs, complex_taps, beam_phases,
                              beam_direction_counts, rx_metadata.initialization_time(),
-                             rx_metadata.sequence_start_time(), slice_ids);
+                             rx_metadata.sequence_start_time(), slice_ids, dm_rates);
 
     if (rx_metadata.numberofreceivesamples() == 0){
       //TODO(keith): handle error for missing number of samples.
     }
 
-    //We need to sample early to account for propagating samples through filters. The number of 
+
+    //We need to sample early to account for propagating samples through filters. The number of
     //required early samples is equal to the largest filter length in time (based on the rate at
-    //that stage.) The following lines find the max filter length in time and convert that to 
+    //that stage.) The following lines find the max filter length in time and convert that to
     //number of samples at the input rate.
-    int64_t extra_samples = (first_stage_dm_rate * second_stage_dm_rate * third_stage_dm_rate *
-                        filters.get_num_fourth_stage_taps());
 
-    if (extra_samples < (first_stage_dm_rate * second_stage_dm_rate * filters.get_num_third_stage_taps())) {
-      extra_samples = first_stage_dm_rate * second_stage_dm_rate * filters.get_num_third_stage_taps();
-    }
+    int64_t extra_samples = std::accumulate(dm_rates.begin(), dm_rates.end()-1, 1,
+                                            std::multiplies<int64_t>()) *
+                            complex_taps.back().size();
 
-    if (extra_samples < (first_stage_dm_rate * filters.get_num_second_stage_taps())) {
-      extra_samples = first_stage_dm_rate * filters.get_num_second_stage_taps();
-    }
+    // problem with this if you have too few stages. Need to figure this out.
 
-    if (extra_samples < (filters.get_num_first_stage_taps())) {
-      extra_samples = filters.get_num_first_stage_taps();
+    for (int32_t i=complex_taps.size()-2; i>=0; i--) {
```

Comment:
This part @mardet987. Currently requires a few stages minimum for this to work.

Reply:
at least two stages I suppose?

Reply:
Let's talk about this. I'm not sure what this for loop is doing.

Reply:
Sorry. I think I get this now. I think it all could be done in a for loop without the ifs though and might make it more readable. Not sure if the following would work if dm_rates.size() = 1 though. But something like this (and removing the earlier line at 240?) 
```suggestion 
    extra_samples = 0
    for (int32_t i=0; i<dm_rates.size(); i++) {
        stage_extra_samples = std::accumulate(dm_rates.begin(), dm_rates[i], 1, std::multiplies<int64_t>()) * complex_taps[i].size()
        if stage_extra_samples > extra_samples:
            extra_samples = stage_extra_samples
```